contents/data/nosql/NoSQL.md

```diff
@@ -0,0 +1,407 @@
+```
+
+Here, the code simply provides a query JSON object matching the structure of an item document.
+
+Specifically, items with a price of less than 30 are filtered with the `$lt: 30` key-value pair,
+and items with a popularity of more than 8 are filtered with the `$gt: 8`.<br>
+Subsequently, a "cursor" is returned, which undergoes a `sort` operation by the respective fields of `price`
+and `title`.
+
+CRUD operations with NoSQL databases are just as powerful as traditional SQL databases, and in some
+cases even more succinct.
+
+<br>
+
+### Schemas in NoSQL
+
+One key characteristic of most NoSQL databases is that they are **schema-less**.
+
+This means that each individual document has no restriction on what keys it must have,
+the number of keys, the type of values, etc.
+(Note the missing fields for the second item in the earlier code example [above](#schemaless), which are intentionally omitted.)
+
+Documents can even contain other documents (json objects), arrays - anything the database used
+can serialize and deserialize.
+
+#### Validation
+
+At the same time, NoSQL databases usually also provide some degree of schema validation.
+
+For example, in the _customers_ collection, where the fields of a customer are unlikely to
+change, it can be helpful to enforce a strict schema on documents to prevent erroneous database
+operations.
+
+```js
+...
+{
+  $jsonSchema: {
+    bsonType: "object",
+	require: [ "username", "password", "email" ]
+	properties: {
+	  ...
+	}
+  }
+}
+...
+```
+
+<br>
+
+### Relation storage in NoSQL
+
+The world is full of relations. For example, a patient _is related to_ his / her disease record,
+just as a customer _is related to_ their shopping cart.
+
+Sometimes, the objects on both sides of the relation can contain substantial amounts of
+information, and may be impossible to store as a singular field in one or the other document.
+
+Hence, simple relations such as one-to-one relations, one-to-many relations are often expressed in NoSQL
+databases directly in the form of embedded documents.
+
+For example, for a customer and his / her shopping cart, we may have the following:
+```js
+{
+  username: "panda",
+  cart: {
+    totalPrice: 100,
+    cartItems: [ ... ],
+    discountCode: "panda"
+  },
+  email: "panda@pandas.com",
+  ...
+}
+```
+
+Such simple relations are extremely suited to and expressable with the schema-less characteristic
+of NoSQL databases.
+
+In the case of more complicated many-to-many relationships, relations are commonly stored using references,
+to avoid duplication of data.
+
+For example, items in an e-commerce website are related to the many customers through their carts.
+In these carts, it is much more space efficient to store references (to the items), than the item documents themselves.
+
+Let's introduce a uniquely generated `_id` field for each item document in the items collection:
+```js
+{
+  type: "book",
+  price: 20,
+  popularity: 9.7,
+  _id: 9d1793bd491349n913847n93d
```

Comment:
Hm I'm also new to nosql, but over here _id is some kind of unquoted string, but later on it's inside quotes ""s.

Are they interchangeably? I was a bit confused at this part.

Reply:
Good catch, thanks!